### Data Cleaning 
- Handling Missing / Null Values
- Removing Duplicate Records
- Fixing Invalid Data Types
- Handling Inconsistent Values
- Correcting Invalid / Impossible Values
- Standardizing Categorical Values
- Detecting & Handling Outliers
- Handling Skewed Distributions
- Correcting Formatting Inconsistencies
- Handling Incorrect Date & Time Formats
- Fixing Data Entry Errors
- Validating Data Ranges & Constraints
- Handling Noisy Data
- Removing Irrelevant / Redundant Data
- Ensuring Data Consistency & Quality

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('../E-commerce_data/EDA_completed.csv')

In [4]:
df.head(3)

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,...,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount
0,0,ORD1048298,CUST55152,2025.08.45,33.0,Lucknow,Grocery,Bluetooth Speaker,1.0,1270.11,...,COD,Delivered,3.6,4.0,4.74,Desktop,Social,10552.73,NEWUSER,1076.16
1,1,ORD1081047,CUST35058,2025-08-14,28.0,Chennai,Home & Kitchen,Running Shoes,1.0,1030.57,...,UPI,Delivered,5.0,3.0,16.72,Mobile,Social,894.36,FESTIVE,1015.11
2,2,ORD1092754,CUST67495,2025-11-08,31.0,Pune,Beauty,Face Serum,1.0,284.83,...,UPI,Delivered,5.0,6.0,4.40,Mobile,Organic,9866.55,NEWUSER,240.25


In [7]:
df.isnull().mean() * 100

Unnamed: 0            0.000000
OrderId               0.000000
CustomerId            0.000000
OrderDate             0.594644
Age                   0.000000
City                  1.259546
Category              1.520212
ProductName           0.397108
Quantity              0.689339
UnitPrice             1.092557
DiscountPct           1.598615
PaymentMethod         1.317585
OrderStatus           0.794216
Rating                2.789940
ShippingDays          3.444659
DeliveryDistanceKM    2.082273
Device                0.912331
MarketingChannel      1.628144
Spend90d              2.593422
CouponCode            4.479177
TotalAmount           0.885857
dtype: float64

### Handling Inconsistent value in Categorical columns

In [8]:
df.head(2)

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,...,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount
0,0,ORD1048298,CUST55152,2025.08.45,33.0,Lucknow,Grocery,Bluetooth Speaker,1.0,1270.11,...,COD,Delivered,3.6,4.0,4.74,Desktop,Social,10552.73,NEWUSER,1076.16
1,1,ORD1081047,CUST35058,2025-08-14,28.0,Chennai,Home & Kitchen,Running Shoes,1.0,1030.57,...,UPI,Delivered,5.0,3.0,16.72,Mobile,Social,894.36,FESTIVE,1015.11


In [ ]:
df['OrderDate'] = df['OrderDate'].str.replace(".","-")


In [18]:
df['OrderDate'] = df['OrderDate'].str.replace("/","-")
df['OrderDate'] = df['OrderDate'].str.replace(","," ")
df['OrderDate'] = df['OrderDate'].str.strip()


In [21]:
df['OrderDate'] = df['OrderDate'].str.replace("not available","nan")

In [25]:
df['OrderDate'] = pd.to_datetime(df['OrderDate'], errors='coerce')

C:\Users\md salman\AppData\Local\Temp\ipykernel_14436\2282615567.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['OrderDate'] = pd.to_datetime(df['OrderDate'], errors='coerce')


In [28]:
df['OrderDate'].unique()

<DatetimeArray>
[                'NaT', '2025-08-14 00:00:00', '2025-11-08 00:00:00',
 '2024-12-01 00:00:00', '2025-07-15 00:00:00', '2025-01-13 00:00:00',
 '2024-11-10 00:00:00', '2025-11-18 00:00:00', '2025-06-28 00:00:00',
 '2024-06-01 00:00:00',
 ...
 '2025-12-16 00:00:00', '2024-11-13 00:00:00', '2024-10-06 00:00:00',
 '2025-08-20 00:00:00', '2025-06-30 00:00:00', '2025-04-04 00:00:00',
 '2024-03-27 00:00:00', '2025-01-24 00:00:00', '2024-10-15 00:00:00',
 '2025-12-31 00:00:00']
Length: 732, dtype: datetime64[ns]

In [29]:
df['Year'] =  df['OrderDate'].dt.year
df['Month'] = df['OrderDate'].dt.month
df['Day'] =  df['OrderDate'].dt.day

In [30]:
df.head(2)

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,...,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
0,0,ORD1048298,CUST55152,NaT,33.0,Lucknow,Grocery,Bluetooth Speaker,1.0,1270.11,...,4.0,4.74,Desktop,Social,10552.73,NEWUSER,1076.16,NaN,NaN,NaN
1,1,ORD1081047,CUST35058,2025-08-14,28.0,Chennai,Home & Kitchen,Running Shoes,1.0,1030.57,...,3.0,16.72,Mobile,Social,894.36,FESTIVE,1015.11,2025.0,8.0,14.0


In [32]:
pd.set_option("display.max_columns", None)

In [33]:
df.head(2)

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,DiscountPct,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
0,0,ORD1048298,CUST55152,NaT,33.0,Lucknow,Grocery,Bluetooth Speaker,1.0,1270.11,15.27,COD,Delivered,3.6,4.0,4.74,Desktop,Social,10552.73,NEWUSER,1076.16,NaN,NaN,NaN
1,1,ORD1081047,CUST35058,2025-08-14,28.0,Chennai,Home & Kitchen,Running Shoes,1.0,1030.57,1.50,UPI,Delivered,5.0,3.0,16.72,Mobile,Social,894.36,FESTIVE,1015.11,2025.0,8.0,14.0


In [35]:
df['Age'].unique()
df['Age'] = df['Age'].astype(int)

In [37]:
df['City'].unique()

array(['Lucknow', 'Chennai', 'Pune', 'Indore', 'Patna', 'Ahmedabad',
       'Bengaluru', 'Mumbai', 'Jaipur', 'Hyderabad', 'Delhi', 'Kolkata',
       nan, 'Delhi ', '?', 'Bombay', 'Bengaluru ', 'not provided',
       ' delhi', 'DELHI', 'Hyd', 'bangalore', 'unknown', 'PUNE', '-'],
      dtype=object)

In [44]:
city_map = {
    'Lucknow':'Lucknow',
    'Chennai':'Chennai',
    'Pune':'Pune',
    'Indore':'Indore',
    'Patna':'Patna',
    'Ahmedabad':'Ahmedabad',
    'Bengaluru':'Bengaluru',
    'Mumbai':'Mumbai',
    'Jaipur':'Jaipur',
    'Hyderabad':'Hyderabad',
    'Delhi':'Delhi',
    'Kolkata':'Kolkata',
    'nan':'nan',
    'Delhi ':'Delhi',
    '?':'not provided',
    'Bombay':'Mumbai',
    'Bengaluru ':'Bengaluru',
    'not provided':'not provided',
    ' delhi':'Delhi',
    'DELHI':'Delhi',
    'Hyd':'Hyderabad',
    'bangalore':'Bengaluru',
    'unknown':'not provided',
    'PUNE':'Pune',
    '-':'not provided'
}
df['City'] = df['City'].map(city_map)

In [46]:
df['City'] = df['City'].str.strip()

In [47]:
df.head(2)

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,DiscountPct,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
0,0,ORD1048298,CUST55152,NaT,33,Lucknow,Grocery,Bluetooth Speaker,1.0,1270.11,15.27,COD,Delivered,3.6,4.0,4.74,Desktop,Social,10552.73,NEWUSER,1076.16,NaN,NaN,NaN
1,1,ORD1081047,CUST35058,2025-08-14,28,Chennai,Home & Kitchen,Running Shoes,1.0,1030.57,1.50,UPI,Delivered,5.0,3.0,16.72,Mobile,Social,894.36,FESTIVE,1015.11,2025.0,8.0,14.0


In [50]:
category_map = {
    'Grocery': 'Grocery',
    'Home & Kitchen': 'Home & Kitchen',
    'Home&Kitchen': 'Home & Kitchen',
    'home & kitchen': 'Home & Kitchen',

    'Beauty': 'Beauty',
    'Electronics': 'Electronics',
    'ELECTRONICS': 'Electronics',
    'electronics': 'Electronics',
    ' Electronics ': 'Electronics',

    'Toys': 'Toys',
    'Books': 'Books',
    'Sports': 'Sports',

    'Fashion': 'Fashion',
    'fashion ': 'Fashion',

    'nan': 'not provided',
    '-': 'not provided',
    '?': 'not provided',
    'unknown': 'not provided',
    'not provided': 'not provided'
}

df['Category'] = df['Category'].map(category_map)

In [57]:
payment_map = {
    'COD': 'COD',
    'COD ': 'COD',
    'cash on delivery': 'COD',

    'UPI': 'UPI',
    'upi': 'UPI',
    'UPI ': 'UPI',

    'Credit Card': 'Credit Card',
    'Credit card': 'Credit Card',
    'credit_card': 'Credit Card',

    'Debit Card': 'Debit Card',

    'Wallet': 'Wallet',

    'Net Banking': 'Net Banking',
    'NetBanking': 'Net Banking',

    '-': 'not provided',
    'unknown': 'not provided',
    'not provided': 'not provided',
    '?': 'not provided',
    '': 'not provided'
}

df['PaymentMethod'] = df['PaymentMethod'].map(payment_map)
df['PaymentMethod'] = df['PaymentMethod'].fillna('not provided')

In [66]:
device_map = {
    'Desktop': 'Desktop',
    'desktop ': 'Desktop',
    'pc': 'Desktop',

    'Mobile': 'Mobile',
    'Mobile ': 'Mobile',
    'mobile': 'Mobile',
    'MOBILE': 'Mobile',

    'Tablet': 'Tablet',
    'Tab': 'Tablet',

    'not provided': 'not provided',
    'unknown': 'not provided',
    '-': 'not provided',
    '?': 'not provided'
}

df['Device'] = df['Device'].map(device_map)
df['Device'] = df['Device'].fillna('not provided')

In [70]:
channel_map = {
    'Social': 'Social',
    'social media': 'Social',

    'Organic': 'Organic',
    'Organic ': 'Organic',
    'organic': 'Organic',

    'Paid Search': 'Paid Search',
    'paid search': 'Paid Search',
    'Paid_Search': 'Paid Search',

    'Email': 'Email',

    'Affiliate': 'Affiliate',

    'Direct': 'Direct',
    'direct ': 'Direct',

    'not provided': 'not provided',
    '?': 'not provided',
    '-': 'not provided',
    'unknown': 'not provided'
}

df['MarketingChannel'] = df['MarketingChannel'].map(channel_map)
df['MarketingChannel'] = df['MarketingChannel'].fillna('not provided')

In [74]:
discount_map = {
    'NEWUSER': 'NEWUSER',
    'FESTIVE': 'FESTIVE',
    'APP15': 'APP15',
    'SAVE20': 'SAVE20',
    'NONE': 'NONE',
    '-': 'not provided',
    '?': 'not provided',
    'not provided': 'not provided',
    'unknown': 'not provided'
}

df['CouponCode'] = df['CouponCode'].map(discount_map)
df['CouponCode'] = df['CouponCode'].fillna('not provided')

### Handing Invalid Values in Numerical  columns

In [76]:
df.head(2)

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,DiscountPct,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
0,0,ORD1048298,CUST55152,NaT,33,Lucknow,Grocery,Bluetooth Speaker,1.0,1270.11,15.27,COD,Delivered,3.6,4.0,4.74,Desktop,Social,10552.73,NEWUSER,1076.16,NaN,NaN,NaN
1,1,ORD1081047,CUST35058,2025-08-14,28,Chennai,Home & Kitchen,Running Shoes,1.0,1030.57,1.50,UPI,Delivered,5.0,3.0,16.72,Mobile,Social,894.36,FESTIVE,1015.11,2025.0,8.0,14.0


In [78]:
df['Age'].unique()

array([ 33,  28,  31,  38,  18,  48,  22,  46,  25,  49,  26,  34,  39,
        32,  44,  29,  41,  45,  43,  54,  42,  21,  -5,  40,  95,  36,
        23,  35, 121,  27,  30,  51,  37,  53,  24,  47,  56,  50,  19,
        61, 104,  20,  57,  59,   0,  69,  52,  55,  62,  58,  96,  60,
       106,  85,  90,  66, 105,  64,  67,  63, 102,  86,  88, 101, 110,
        87,  91, 150, 108,  65,  99,  92,  89, 103,  94,  93, 107, 100,
        97,  98,  68, 109,  70])

In [81]:
df = df[df['Quantity'] >= 0]

In [87]:
df[df['UnitPrice'] < 0]

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,DiscountPct,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
63,63,ORD1012442,CUST74068,2024-03-31,49,Kolkata,Grocery,Organic Snacks,1.0,-50.0,15.48,Net Banking,Shipped,5.0,4.0,31.41,Mobile,Social,183022.70,NEWUSER,164.67,2024.0,3.0,31.0
2168,2204,ORD1044808,CUST31717,2024-11-23,32,Indore,Beauty,Smart Watch,1.0,-999.0,4.66,UPI,Delivered,3.9,2.0,5.04,Mobile,not provided,5722.46,not provided,1146.82,2024.0,11.0,23.0
2820,2867,ORD1085461,CUST96847,2025-09-15,42,Jaipur,Grocery,Smart Watch,2.0,-999.0,20.04,Net Banking,Delivered,3.7,8.0,5.10,Desktop,Paid Search,3291.56,SAVE20,5219.01,2025.0,9.0,15.0
2937,2985,ORD1033481,CUST86591,2024-09-01,58,Pune,Electronics,Bluetooth Speaker,5.0,-999.0,12.26,COD,Delivered,2.8,2.0,4.06,Tablet,Organic,7016.59,NONE,2231.14,2024.0,9.0,1.0
2955,3003,ORD1014732,CUST98581,2024-04-17,56,Indore,Electronics,Face Serum,1.0,-999.0,6.02,COD,Delivered,4.7,7.0,6.21,Mobile,not provided,1465.34,SAVE20,178.80,2024.0,4.0,17.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95323,97054,ORD1041408,CUST99078,2024-10-29,36,Bengaluru,Fashion,Yoga Mat,1.0,-50.0,9.37,COD,Delivered,4.1,4.0,9.22,Mobile,Social,513.74,APP15,2535.66,2024.0,10.0,29.0
95478,97215,ORD1076105,CUST98596,2025-07-09,38,Chennai,Home & Kitchen,Running Shoes,1.0,-50.0,1.11,Net Banking,Cancelled,3.3,NaN,88.94,Mobile,Organic,5623.30,SAVE20,1684.20,2025.0,7.0,9.0
95574,97311,ORD1053656,CUST96955,2025-01-26,35,Pune,Toys,Novel,1.0,-999.0,22.81,UPI,Delivered,4.2,6.0,8.71,Desktop,Paid Search,1025.38,APP15,681.97,2025.0,1.0,26.0
96561,98318,ORD1095242,CUST90709,2025-11-26,29,Hyderabad,Books,Coffee Maker,1.0,-50.0,20.55,UPI,Shipped,4.3,2.0,49.60,Desktop,Paid Search,1697.49,SAVE20,161.67,2025.0,11.0,26.0


In [88]:
df.loc[df['UnitPrice'] < 0, 'UnitPrice'].value_counts()

UnitPrice
-999.0    96
-50.0     88
Name: count, dtype: int64

In [89]:
df.loc[df['UnitPrice'] < 0, ['Category', 'ProductName', 'Quantity', 'UnitPrice', 'TotalAmount']]

,Category,ProductName,Quantity,UnitPrice,TotalAmount
63,Grocery,Organic Snacks,1.0,-50.0,164.67
2168,Beauty,Smart Watch,1.0,-999.0,1146.82
2820,Grocery,Smart Watch,2.0,-999.0,5219.01
2937,Electronics,Bluetooth Speaker,5.0,-999.0,2231.14
2955,Electronics,Face Serum,1.0,-999.0,178.80
...,...,...,...,...,...
95323,Fashion,Yoga Mat,1.0,-50.0,2535.66
95478,Home & Kitchen,Running Shoes,1.0,-50.0,1684.20
95574,Toys,Novel,1.0,-999.0,681.97
96561,Books,Coffee Maker,1.0,-50.0,161.67


In [92]:
df.loc[df['UnitPrice'] < 0, 'TotalAmount'].describe()

count    0.0
mean     NaN
std      NaN
min      NaN
25%      NaN
50%      NaN
75%      NaN
max      NaN
Name: TotalAmount, dtype: float64

In [93]:
df = df[df['UnitPrice'] >= 0]

### Observation

* The `UnitPrice` column contains **184 negative values**.
* The negative values are mainly **-50 and -999**, indicating specific invalid values rather than normal price variation.
* Negative `UnitPrice` values occur across **different categories and products**, so the issue is not limited to a particular category or product.
* `Quantity` values remain positive in the observed records.
* `TotalAmount` remains positive, with a minimum value of **97.48**, despite the negative `UnitPrice`.
* One record has a missing `TotalAmount`, resulting in **183 non-null TotalAmount values** among the 184 affected rows.
* Since a product's unit price should not normally be negative, these records can be treated as **invalid observations**.
* No reliable information is available to determine the correct positive price for these records, so replacing `-50` or `-999` with arbitrary


In [94]:
df.head(2)

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,DiscountPct,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
0,0,ORD1048298,CUST55152,NaT,33,Lucknow,Grocery,Bluetooth Speaker,1.0,1270.11,15.27,COD,Delivered,3.6,4.0,4.74,Desktop,Social,10552.73,NEWUSER,1076.16,NaN,NaN,NaN
1,1,ORD1081047,CUST35058,2025-08-14,28,Chennai,Home & Kitchen,Running Shoes,1.0,1030.57,1.50,UPI,Delivered,5.0,3.0,16.72,Mobile,Social,894.36,FESTIVE,1015.11,2025.0,8.0,14.0


In [96]:
df['DiscountPct'].value_counts()

DiscountPct
 101.00    77
 999.00    76
-10.00     74
 8.02      70
 5.19      70
           ..
 40.75      1
 41.45      1
 38.88      1
 35.05      1
 38.73      1
Name: count, Length: 4083, dtype: int64

In [97]:
df[df['DiscountPct'] < 0]

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,DiscountPct,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
234,236,ORD1041440,CUST93245,2024-10-29,27,Lucknow,Electronics,Organic Snacks,1.0,573.26,-10.0,COD,Shipped,3.8,NaN,NaN,Desktop,Email,10276.15,not provided,539.32,2024.0,10.0,29.0
2504,2545,ORD1098718,CUST89409,2025-12-21,31,Indore,Home & Kitchen,Laptop Stand,1.0,3411.25,-10.0,UPI,Delivered,3.7,3.0,5.61,Mobile,Direct,2844.35,APP15,3383.96,2025.0,12.0,21.0
3316,3367,ORD1004938,CUST79378,2024-02-06,39,Patna,Beauty,Bluetooth Speaker,2.0,1403.25,-10.0,Credit Card,Delivered,4.0,5.0,2.76,Tablet,Direct,4927.74,NEWUSER,2359.42,2024.0,2.0,6.0
4781,4871,ORD1039920,CUST57482,2024-10-18,34,Delhi,Fashion,Wireless Headphones,1.0,460.68,-10.0,Net Banking,Delivered,NaN,6.0,8.37,Mobile,Affiliate,11567.79,APP15,416.22,2024.0,10.0,18.0
5344,5443,ORD1075044,CUST81808,2025-07-01,28,Jaipur,not provided,Coffee Maker,4.0,243.55,-10.0,Wallet,Delivered,4.8,2.0,21.41,Mobile,not provided,957.86,NEWUSER,846.58,2025.0,7.0,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
89374,90999,ORD1041553,CUST27463,2024-10-30,39,Indore,Grocery,Wireless Headphones,3.0,173.98,-10.0,Wallet,not provided,4.4,7.0,15.02,Mobile,Paid Search,5740.02,FESTIVE,441.09,2024.0,10.0,30.0
90227,91870,ORD1026676,CUST67227,2024-07-13,39,Mumbai,Fashion,Organic Snacks,1.0,588.57,-10.0,UPI,Delivered,4.6,4.0,10.96,Mobile,Paid Search,3152.70,SAVE20,417.24,2024.0,7.0,13.0
93720,95427,ORD1080231,CUST62641,2025-08-08,32,Bengaluru,Electronics,Organic Snacks,1.0,2373.40,-10.0,Net Banking,Delivered,3.8,1.0,10.70,Mobile,Organic,19315.95,NEWUSER,2163.59,2025.0,8.0,8.0
94348,96064,ORD1004381,CUST91186,2024-02-01,35,Bengaluru,Electronics,Novel,1.0,317.05,-10.0,Net Banking,Delivered,4.0,7.0,9.78,Desktop,Organic,9171.35,SAVE20,278.75,2024.0,2.0,1.0


In [98]:
df.loc[df['DiscountPct'] < 0, 'DiscountPct'].value_counts()


DiscountPct
-10.0    74
Name: count, dtype: int64

In [99]:
df.loc[df['DiscountPct'] < 0, ['Category', 'ProductName', 'Quantity', 'UnitPrice', 'TotalAmount']]


,Category,ProductName,Quantity,UnitPrice,TotalAmount
234,Electronics,Organic Snacks,1.0,573.26,539.32
2504,Home & Kitchen,Laptop Stand,1.0,3411.25,3383.96
3316,Beauty,Bluetooth Speaker,2.0,1403.25,2359.42
4781,Fashion,Wireless Headphones,1.0,460.68,416.22
5344,not provided,Coffee Maker,4.0,243.55,846.58
...,...,...,...,...,...
89374,Grocery,Wireless Headphones,3.0,173.98,441.09
90227,Fashion,Organic Snacks,1.0,588.57,417.24
93720,Electronics,Organic Snacks,1.0,2373.40,2163.59
94348,Electronics,Novel,1.0,317.05,278.75


### Observation

* The `DiscountPct` column contains **74 negative values**.
* All 74 invalid values are exactly **-10.00%**.
* A discount percentage should not be negative in this dataset, so `-10.00%` represents an **invalid value**.
* The invalid values occur across different categories, products, payment methods, and order statuses, indicating that the issue is **not restricted to a particular category or product**.
* The corresponding `UnitPrice` and `TotalAmount` values are positive, while only `DiscountPct` is negative.
* Other suspicious values such as **77 and 999** also appear frequently in the column and should be investigated because they may be outside the expected discount range.
* Since there is no reliable information to determine the intended discount for the 74 affected records, replacing `-10.00` with an arbitrary value would not be appropriate.
* Therefore, the negative discount values should be treated as **invalid data and handled during the data-cleaning stage**.


In [100]:
df = df[df['DiscountPct'] >= 0]

In [103]:
df = df[df['Rating'] >= 0]

In [104]:
df.head(2)

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,DiscountPct,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
0,0,ORD1048298,CUST55152,NaT,33,Lucknow,Grocery,Bluetooth Speaker,1.0,1270.11,15.27,COD,Delivered,3.6,4.0,4.74,Desktop,Social,10552.73,NEWUSER,1076.16,NaN,NaN,NaN
1,1,ORD1081047,CUST35058,2025-08-14,28,Chennai,Home & Kitchen,Running Shoes,1.0,1030.57,1.50,UPI,Delivered,5.0,3.0,16.72,Mobile,Social,894.36,FESTIVE,1015.11,2025.0,8.0,14.0


In [106]:
df[df['ShippingDays'] < 0]

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,DiscountPct,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
86,86,ORD1053983,CUST56066,2025-01-29,121,Patna,Books,Organic Snacks,5.0,658.61,13.07,Wallet,Delivered,3.8,-2.0,3.23,Mobile,Paid Search,1620.06,APP15,2862.65,2025.0,1.0,29.0
536,541,ORD1062275,CUST51493,2025-03-30,24,Ahmedabad,Toys,Novel,1.0,612.61,29.23,Net Banking,Delivered,4.3,-2.0,14.81,Mobile,Paid Search,3388.63,APP15,433.54,2025.0,3.0,30.0
627,632,ORD1032040,CUST18587,2024-08-21,51,not provided,Sports,Coffee Maker,2.0,468.57,14.28,COD,Shipped,4.6,-4.0,11.53,Mobile,Organic,6460.55,NEWUSER,803.32,2024.0,8.0,21.0
1272,1292,ORD1076338,CUST59275,2025-07-11,38,Mumbai,Fashion,Laptop Stand,1.0,736.23,18.37,UPI,Processing,3.9,-4.0,9.47,Desktop,Affiliate,3463.84,FESTIVE,600.98,2025.0,7.0,11.0
1785,1817,ORD1072377,CUST50501,2025-06-12,33,Chennai,Beauty,Wireless Headphones,1.0,91.90,11.44,Wallet,Delivered,4.4,-2.0,5.41,Mobile,Social,664.11,FESTIVE,81.39,2025.0,6.0,12.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
96040,97788,ORD1055401,CUST30594,2025-02-08,23,Mumbai,Sports,Organic Snacks,4.0,479.59,7.78,COD,Returned,3.5,-4.0,11.10,Mobile,Affiliate,8987.70,not provided,1769.11,2025.0,2.0,8.0
96387,98140,ORD1091107,CUST89576,2025-10-27,32,Indore,Toys,Face Serum,2.0,674.41,9.42,COD,Shipped,4.9,-4.0,4.27,Mobile,Social,2588.22,SAVE20,1221.76,2025.0,10.0,27.0
96681,98441,ORD1015997,CUST83665,2024-04-26,39,Kolkata,Fashion,Face Serum,1.0,336.41,10.05,Credit Card,Delivered,3.1,-2.0,6.06,Desktop,Direct,4450.78,FESTIVE,302.60,2024.0,4.0,26.0
96750,98511,ORD1069489,CUST88319,2025-05-22,35,Jaipur,Electronics,Coffee Maker,1.0,450.41,5.06,Credit Card,Delivered,4.9,-2.0,24.47,not provided,Organic,7653.72,NEWUSER,427.62,2025.0,5.0,22.0


In [107]:
df = df[df['ShippingDays'] >= 0]

In [110]:
df['DeliveryDistanceKM'].value_counts()

DeliveryDistanceKM
100.00     204
0.00        99
9999.00     88
4.12        86
4.60        86
          ... 
1329.00      1
4228.00      1
47.00        1
58.22        1
63.85        1
Name: count, Length: 6348, dtype: int64

In [111]:
df.head(2)

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,DiscountPct,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
0,0,ORD1048298,CUST55152,NaT,33,Lucknow,Grocery,Bluetooth Speaker,1.0,1270.11,15.27,COD,Delivered,3.6,4.0,4.74,Desktop,Social,10552.73,NEWUSER,1076.16,NaN,NaN,NaN
1,1,ORD1081047,CUST35058,2025-08-14,28,Chennai,Home & Kitchen,Running Shoes,1.0,1030.57,1.50,UPI,Delivered,5.0,3.0,16.72,Mobile,Social,894.36,FESTIVE,1015.11,2025.0,8.0,14.0


Spend90d
1321.76    4
4032.33    4
2787.98    3
1838.74    3
1911.43    3
          ..
6290.70    1
2237.47    1
740.05     1
300.35     1
2021.02    1
Name: count, Length: 83519, dtype: int64

### Handling Outliers

In [114]:
# Columns: Age
# Outliers: 917

# Columns: Quantity
# Outliers: 10302

# Columns: DiscountPct
# Outliers: 1680

# Columns: Rating
# Outliers: 710

# Columns: ShippingDays
# Outliers: 4285

# Columns: DeliveryDistanceKM
# Outliers: 6733

# Columns: Spend90d
# # Outliers: 8584

In [115]:
df['Age'].unique()

array([ 33,  28,  31,  18,  48,  22,  46,  38,  25,  49,  26,  34,  39,
        32,  44,  29,  41,  45,  43,  54,  42,  21,  -5,  40,  95,  36,
        23,  35,  27,  30,  51,  37,  53,  24,  47,  56,  50,  19,  61,
       104,  20,  57, 121,  59,   0,  69,  52,  55,  62,  58,  96,  60,
       106,  85,  90,  66, 105,  64,  63, 102,  86,  88, 101, 110,  87,
        91, 150, 108,  65,  99,  92,  67,  89, 103,  94,  93, 107, 100,
        97,  98,  68, 109,  70])

In [116]:
df.query('Age > 100')

,Unnamed: 0,OrderId,CustomerId,OrderDate,Age,City,Category,ProductName,Quantity,UnitPrice,DiscountPct,PaymentMethod,OrderStatus,Rating,ShippingDays,DeliveryDistanceKM,Device,MarketingChannel,Spend90d,CouponCode,TotalAmount,Year,Month,Day
230,232,ORD1046767,CUST36427,2024-12-07,104,Jaipur,Electronics,Coffee Maker,1.0,2893.10,9.57,COD,Delivered,4.0,2.0,4.54,Mobile,Direct,5514.99,not provided,2616.23,2024.0,12.0,7.0
294,296,ORD1038470,CUST35443,2024-10-07,121,Chennai,Beauty,Wireless Headphones,1.0,11968.34,6.66,Debit Card,Delivered,3.6,3.0,1.14,Tablet,Paid Search,4294.68,not provided,11171.25,2024.0,10.0,7.0
1420,1445,ORD1009358,CUST76841,2024-03-09,104,Lucknow,Electronics,Bluetooth Speaker,2.0,95.49,12.82,Net Banking,Delivered,3.0,3.0,9.24,Desktop,Organic,4048.30,SAVE20,166.50,2024.0,3.0,9.0
1600,1628,ORD1053506,CUST53443,2025-01-25,121,Delhi,Beauty,Smart Watch,2.0,610.96,16.17,Credit Card,Returned,4.5,2.0,7.07,Mobile,Email,12185.20,not provided,1024.34,2025.0,1.0,25.0
1736,1767,ORD1005026,CUST89534,2024-02-06,106,Patna,Electronics,Coffee Maker,1.0,3278.64,15.55,COD,Delivered,4.5,1.0,7.42,Desktop,Email,650.50,APP15,2768.81,2024.0,2.0,6.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95148,96875,ORD1077012,CUST68578,2025-07-16,102,Patna,Fashion,Yoga Mat,1.0,413.56,17.31,UPI,Shipped,4.4,2.0,6.45,Mobile,Paid Search,30671.35,APP15,341.97,2025.0,7.0,16.0
95182,96910,ORD1017703,CUST11428,2024-05-09,102,Indore,Beauty,Smart Watch,1.0,4187.23,10.51,Wallet,Delivered,4.5,2.0,3.32,Desktop,Direct,3337.63,APP15,3747.15,2024.0,5.0,9.0
95587,97326,ORD1085471,CUST32099,2025-09-15,108,Jaipur,Sports,Face Serum,3.0,343.61,19.89,UPI,Shipped,4.2,1.0,5.57,Mobile,Paid Search,3101.14,SAVE20,825.80,2025.0,9.0,15.0
96503,98259,ORD1088696,CUST94658,2025-10-09,103,Delhi,Beauty,Face Serum,1.0,1137.18,4.82,Net Banking,Shipped,3.7,4.0,11.34,Tablet,Direct,4599.84,not provided,1082.37,2025.0,10.0,9.0


In [117]:
df  = df[df['Age'] < 100]

In [119]:
df.to_csv("Cleaned.csv")

### Counclusion:
In this data cleaning I am hand done inconsistent value outliers invalid values And many words and the data set this time the look like a clean and a clean data we can use the data in feature engineering